Sprint 7 solution in file manager

vinnipooha opened the pull request.

vinnipooha committed:
TS6_fix_2:  устранить ошибки в тестах InMemoryHistoryManagerTest. Исправлен .gitignore, удалены служебные файлы

vinnipooha committed:
TS7: реализовать FileBackedTaskManager.

реализовано тестирование нового менеджера, переопределен метод Managers.getDefault(), исправлен .gitignore, удалены служебные файлы, пользовательский сценарий реализован в классе Main

vinnipooha committed:
TS7: FIX_проведен рефакторинг кода:
1. в классе Status заменен метод toString() на getStatus());
2. ArrayList заменен на List
3. Оптимизированы импорты

andrew-larkin commented on src/exceptions/ManagerSaveException.java:
```diff
@@ -0,0 +1,11 @@
+package exceptions;
+
+public class ManagerSaveException extends RuntimeException {
```
Отлично!

andrew-larkin commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
```
имплементировать TaskManager не требуется, т.к. это уже делает InMemoryTaskManager, от которого мы наследуемся

andrew-larkin commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+    protected Path path;
+    static final String HEADER = "id,type,name,status,description,epic\n";
+
+    public FileBackedTaskManager(HistoryManager historyManager, Path path) {
```
HistoryManager абсолютно точно не нужно передавать в параметрах. Он и так доступен нам из InMemoryTaskManager, просто убедись, что у него там модификатор доступа protected

andrew-larkin commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        this.path = path;
+    }
+
+    public void save() {
```
метод используется только в этом классе, поэтому должен быть приватным

andrew-larkin commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        return fileBakedManager;
+    }
+
+    public void load() {
```
private, т.к. используется только в этом классе

andrew-larkin commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        }
+    }
+
+    public Task taskFromString(String value) throws RuntimeException {
```
эти методы парсинга стоит сделать статическими и выделить в отдельный класс, например, CSVFormat. Это разгрузит текущий класс менеджера от доп. логики

andrew-larkin commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        super.deleteAllSubTasks();
+        save();
+    }
+}
```
в этом классе также должен быть реализован тестовый скрипт. я скину тебе свой. названия могут немного не совпадать - не обращай внимание

public static void main(String[] args) {
   FileBackedTasksManager fileManager = new FileBackedTasksManager(new File("saveTasks2.csv"));
   fileManager.createTask(new Task("task1", "Купить автомобиль"));
   fileManager.createEpic(new Epic("new Epic1", "Новый Эпик"));
   fileManager.createSubtask(new Subtask("New Subtask", "Подзадача", 2));
   fileManager.createSubtask(new Subtask("New Subtask2", "Подзадача2", 2));
   fileManager.getTask(1);
   fileManager.getEpic(2);
   fileManager.getSubtask(3);
   System.out.println(fileManager.getTasks());
   System.out.println(fileManager.getEpics());
   System.out.println(fileManager.getSubTasks());
   System.out.println(fileManager.getHistory());


   System.out.println("\n\n" + "new" + "\n\n");

   FileBackedTasksManager fileBackedTasksManager = loadFromFile(new File("saveTasks2.csv"));
   System.out.println(fileBackedTasksManager.getTasks());
   System.out.println(fileBackedTasksManager.getEpics());
   System.out.println(fileBackedTasksManager.getSubTasks());
   System.out.println(fileBackedTasksManager.getHistory());
}

andrew-larkin commented on src/model/Task.java:
```diff
@@ -60,9 +60,13 @@ public int getId() {
         return id;
     }
 
+    public Type getType() {
```
Type должен быть внедрен в Task в виде нового поля. И также, как и TaskStatus, должен быть проинициализирован в конструкторе (также во всех наследниках Task)

andrew-larkin reviewed (changes requested):
Привет! Очень хорошее решение, все сделано по уму. Остается местами скоррекировать архитектуру и модификаторы доступа и работа будет принята. Тестовый скрипт должен возвращать идентичные распечатки (см коммент в FileBackedTaskManager)

vinnipooha committed:
TS7: FIX_Исправить ошибки по ревью_1:
1. Заведено поле Type и проинициализировано в конструкторе
2. Реализован тестовый скрипт в классе FileBackedTaskManager
3. Логика по преобразованию данных вынесена в отдельный класс CSVFormat
4. Методы save() и load() сделаны приватными
5. HistoryManager удален из параметров конструктора FileBackedTaskManager
6. Внесены правки в тесты, учитывающие внесенные выше изменения
7. "implements TaskManager" из заголовка FileBackedTaskManager удалено
